src/operator/numpy/linalg/np_gesvd-inl.h

```diff
@@ -0,0 +1,298 @@
+
+struct GesvdVecSign {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, int m, int n, DType* UT,
+                                  DType* V, int ldut, int ldv) {
+    DType* vrow(V + i * ldv);
+    DType maxval(fabs(vrow[0])), vval(0.0);
+    int maxind(0);
+    for (int i = 1; i < n; ++i) {
+      vval = fabs(vrow[i]);
+      if (vval > maxval) {
+        maxval = vval;
+        maxind = i;
+      }
+    }
+    if (vrow[maxind] < 0) {
+      DType* utcol(UT + i % m + (i / m) * ldut * m);
+      for (int i = 0; i < n; ++i) {
+        vrow[i] = -vrow[i];
+        if (i < m) {
+          utcol[i * ldut] = -utcol[i * ldut];
+        }
+      }
+    }
+  }
+};
+
+// (UT, L, V) = gesvd(A) [singular value decomposition]
+// - V can overwrite A
+// - Needs workspace (both DType and int), size of which is determined by a
+//   workspace query
+struct gesvd {
+  template<typename xpu, typename DType>
+  static void op(const Tensor<xpu, 3, DType>& A,
+                 const Tensor<xpu, 3, DType>& UT,
+                 const Tensor<xpu, 2, DType>& L,
+                 const Tensor<xpu, 3, DType>& V,
+                 const OpContext& ctx,
+                 const nnvm::NodeAttrs& attrs) {
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    if (A.dptr_ != V.dptr_) Copy(V, A, s);
+    // From here on, we work on V only
+    // Reserve workspace (size determined by query)
+    int lwork(linalg_gesvd_workspace_query(UT[0], L[0], V[0], s));
+    Tensor<xpu, 1, DType> work = ctx.requested[0]
+      .get_space_typed<xpu, 1, DType>(Shape1(lwork), s);
+    // Loop over items in batch
+    for (index_t i = 0; i < UT.size(0); ++i) {
+      linalg_gesvd(UT[i], L[i], V[i], work, s);
+    }
+    // Set signs in a deterministic way
+    using namespace mxnet_op;
+    Kernel<GesvdVecSign, xpu>::Launch
+      (s, V.size(0) * V.size(1), V.size(1), V.size(2),
+       UT.dptr_, V.dptr_, UT.stride_, V.stride_);
+  }
+};
+
+// (A) => (UT, L, V)
+template<typename xpu, typename laop>
+void NumpyLaGesvdForward(const nnvm::NodeAttrs& attrs,
+                         const OpContext& ctx,
+                         const std::vector<TBlob>& inputs,
+                         const std::vector<OpReqType>& req,
+                         const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  CHECK_EQ(inputs.size(), 1);
+  CHECK_EQ(outputs.size(), 3);
+  if (inputs[0].shape_.Size() == 0) {
+    return;
+  }
+  MSHADOW_SGL_DBL_TYPE_SWITCH(outputs[0].type_flag_, OType, {
+    mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+    laop::op(inputs[0].FlatToKD<xpu, 3, OType>(s),
+             outputs[0].FlatToKD<xpu, 3, OType>(s),
+             outputs[1].FlatToKD<xpu, 2, OType>(s),
+             outputs[2].FlatToKD<xpu, 3, OType>(s), ctx, attrs);
+  });
+}
+
+// Helper for gesvd_backward. See technical report for details
+template<typename DType>
+DType gesvd_back_helper_eps(DType* X);
+
+template<>
+MSHADOW_XINLINE float gesvd_back_helper_eps(float* X) {
```

Comment:
Would be cleaner to make this constants dependent on values in std::numeric_limits. So something along the lines of 
std::numeric_limits<DType>::epsilon*10 (or whatever number you need to be on the safe side)
That also avoids doing template specialization here.

Reply:
I tried it out. But I found std::numeric_limits::epsilon() cannot be accessed in Cuda. So I stick with the original implementation for now.